d2/src/main/java/com/linkedin/d2/balancer/dualread/UriPropertiesDualReadMonitor.java

```diff
@@ -0,0 +1,200 @@
+
+      cluster._uris = cluster._oldLb.Uris().size();
+      Set<URI> newLbUris = new HashSet<>(cluster._newLb.Uris());
+
+      for (URI uri : cluster._oldLb.Uris()) {
+        if (!newLbUris.remove(uri)) {
+          continue;
+        }
+
+        if (compareURI(uri, cluster._oldLb, cluster._newLb)) {
+          cluster._matched++;
+        }
+      }
+      // add the remaining unmatched URIs in newLbUris to the uri count
+      cluster._uris += newLbUris.size();
+
+      if (cluster._matched != cluster._uris) {
+        RATE_LIMITED_LOGGER.info("Mismatched cluster properties for {} (match score: {}, total uris: {}):"
+                + "\nOld LB: {}\nNew LB: {}",
+            clusterName, cluster._matched / cluster._uris, cluster._uris, cluster._oldLb, cluster._newLb);
+      }
+
+      _totalUris += cluster._uris;
+      _matchedUris += cluster._matched;
+
+      updateJmxMetrics(clusterName, cluster);
+    }
+  }
+
+  private void updateJmxMetrics(String clusterName, ClusterMatchRecord cluster) {
+    // set a clone of cluster match record to jmx to avoid jmx reading the record in the middle of an update
+    _dualReadLoadBalancerJmx.setClusterMatchRecord(clusterName, cluster == null ? null : cluster.clone());
+    _dualReadLoadBalancerJmx.setUriPropertiesSimilarity((double) _matchedUris / (double) _totalUris);
+  }
+
+  private static boolean compareURI(URI uri, UriProperties oldLb, UriProperties newLb) {
+    String clusterName = oldLb.getClusterName();
+    return compareMaps("partition desc", clusterName, uri, UriProperties::getPartitionDesc, oldLb, newLb) &&
+        compareMaps("specific properties", clusterName, uri, UriProperties::getUriSpecificProperties, oldLb, newLb);
+  }
+
+  private static <K, V> boolean compareMaps(
+      String type, String cluster, URI uri, Function<UriProperties, Map<URI, Map<K, V>>> extractor,
+      UriProperties oldLb, UriProperties newLb
+  ) {
+    Map<K, V> oldData = extractor.apply(oldLb).get(uri);
+    Map<K, V> newData = extractor.apply(newLb).get(uri);
+    if (Objects.equals(oldData, newData)) {
+      return true;
+    }
+
+    LOG.debug("URI {} for {}/{} mismatched between old and new LB.\nOld LB: {}\nNew LB: {}",
+        type, cluster, uri, oldData, newData);
+    return false;
+  }
+
+  @VisibleForTesting
+  int getTotalUris() {
+    return _totalUris;
+  }
+
+  @VisibleForTesting
+  int getMatchedUris() {
+    return _matchedUris;
+  }
+
+  public static class ClusterMatchRecord implements Cloneable {
+    @Nullable
+    @VisibleForTesting
+    UriProperties _oldLb;
+
+    @Nullable
+    @VisibleForTesting
+    UriProperties _newLb;
+
+    @VisibleForTesting
+    int _uris;
+
+    @VisibleForTesting
+    int _matched;
+
+
+
+    public ClusterMatchRecord clone() {
+      try {
+        return (ClusterMatchRecord) super.clone();
+      } catch (CloneNotSupportedException e) {
+        LOG.debug("Failed to clone: {}", this, e);
+        return new ClusterMatchRecord();
+      }
+    }
```

Comment:
See above, can just be:
```suggestion
    public synchronized ClusterMatchRecord clone() {
      return new ClusterMatchRecord(_oldLb, _newLb, _uris, _matched);
    }
```

Reply:
since clone conflicts with the super class method, and IDE complains about not calling super class's clone, I just created a copy method instead

Reply:
Sounds good to me 👍